components-api/src/main/scala/pl/touk/nussknacker/engine/api/definition/ParameterValidator.scala

```diff
@@ -38,12 +42,48 @@ trait Validator {
 
 //TODO: These validators should be moved to separated module
 
+case class SpelExpressionParameterValidator(
+    validationExpression: String,
+    validationFailedMessage: Option[String]
+) extends ParameterValidator {
+
+  private val parser: ExpressionParser   = new SpelExpressionParser()
```

Comment:
It is definitely not a good approach to use spell classes directly. We have some customizations of spel that we want to keep consistent across all spel expressions, we have global helpers as well.
Also, these things should be done outside of the constructor. The class should have objects that already can evaluate the expression.

Reply:
> The class should have objects that already can evaluate the expression.

This little annotation:
`@ConfiguredJsonCodec sealed trait ParameterValidator extends Validator`
makes it very difficult to pass complex objects as constructor parameters. This issue can be by-passed if we implement `CustomParameterValidator` trait, but it still will be impossible to define this validator in config file.

Reply:
> will be impossible to define this validator in config file.

Validators config is IMO the feature that we should consider to remove eventually. After adding `AdditionalUIConfigProvider` there is an easier way to achieve configuratiblity programmatically. I don't know usages where an administrator did this in config manually.

Regards that fact, I see two options:
- `CustomParameterValidator` as you suggested + comment that we don't support it in config - only in `AdditionalUIConfigProvider`
- Write serializer from scratch instead of using generated - it would be difficult to pass some additional context of variables to them, but maybe circe has some options to do that

Reply:
CustomParameterValidator solution does not work, so I finally decided to write deserializer that throws an exception when trying to evaluate an expression.